Support Values in compiled runtime - part 2

henriknyman opened the pull request:
This is based on https://github.com/neo4j/neo4j/pull/11378 with additional commits from "Rework CompiledMaterializeValueMapper" onward.

- Introduce tracking of ValueType and AnyValueType as representation types in CodeGenType
- Let parameters stay as type AnyValue
- Let property values stay as type Value
- Let literal lists and maps be AnyValue
- Handle AnyValue/Value types internally in a bunch of places
- Use the same conversion to storable type as the interpreted runtime, which handles temporal and spatial types

NOTE: This is needed for the compiled runtime to properly handle the new temporal and spatial indexes.

henriknyman committed:
Rework CompiledMaterializeValueMapper

- Remove usage of DefaultValueMapper
- First do a dry run to determine if any conversion is needed,
and if it is not, then return the input value object as it is.

henriknyman committed:
Remove NodeIdWrapper and RelationshipIdWrapper

Replaced by NodeReference and RelationshipReference from
the Values module.

henriknyman committed:
Fix issue with MapExpression

A map expression that we do not know the type of at compile time
can also be a node or a relationship. If those are materialized
values we can get the property directly on them.

- Declare correct type of NodeExpression
- Support boolean variable as predicate

henriknyman committed:
Minor improvement

- Add materializeAnyValueResult
- Replace unused type conversion with assert

henriknyman committed:
Use Java String compare in StringValue

henriknyman committed:
Fix test assertion for utf8 string comparisons

henriknyman committed:
Fix bug in generated hashable tuple equals for primitives

Primitive nodes and relationships are not boxed to Long, which
caused byte code generation to fail.
Generate the equality check for primitive types as an equal
expression instead (a == b).

pontusmelke commented on community/values/src/test/java/org/neo4j/values/storable/UTF8StringValueTest.java:
```diff
@@ -116,10 +116,10 @@ public void shouldCompareTo()
                 int y = utf8Value( string1.getBytes( UTF_8 ) ).compareTo( stringValue( string2 ) );
                 int z = utf8Value( string1.getBytes( UTF_8 ) )
-                        .compareTo( utf8Value( string2.getBytes( UTF_8 ) ) );
+                         .compareTo( utf8Value( string2.getBytes( UTF_8 ) ) );
```
You probably need to update `TextValueSpecification` as well.

henriknyman committed:
Fix bug in compiled aggregation key hashCode and equals

We need to handle primitive types when generating `hashCode`
and `equals` method for aggregation keys with more than one
element (`HashableTupleType`)

henriknyman committed:
Specialize compiled AnyValue[] hashCode

Utilize that an AnyValue cannot be null

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java:
```diff
@@ -45,141 +41,202 @@
+        @Override
+        public AnyValue mapNode( VirtualNodeValue value )
         {
+            if ( value instanceof NodeValue )
```
Maybe
```
{
    needsConversion = !value && value instanceof NodeReference;
    return value;
}
```

henriknyman replied:
I think you mean `needsConversion || (value instanceof NodeReference)`

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java:
```diff
@@ -45,141 +41,202 @@
+        @Override
+        public AnyValue mapRelationship( VirtualRelationshipValue value )
         {
+            if ( value instanceof RelationshipValue )
```
Maybe
```
{
    needsConversion = !needsConversion && value instanceof RelationshipReference;
    return value;
}
```

pontusmelke commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledEquivalenceUtils.java:
```diff
@@ -262,6 +266,29 @@ public static int hashCode( Object[] array )
+     * @return the hash code of the given value
+     */
+    public static int hashCode( AnyValue[] array )
+    {
```
Can't we just use

```
Arrays.hashCode( array );
```

I don't like this trick.

henriknyman replied:
Only for this method or do you mean change all of the hashCode methods that use this?

henriknyman committed:
Fix review comments

henriknyman committed:
Compare UTF8 string values byte by byte
